### PR TITLE
Improve JSON schema validation failure messages

### PR DESCRIPTION
**Pull request by imjoehaines**
## Goal

Use `JSONSchemer::Errors.pretty` to output a nicer JSON schema validation failure message:

- https://www.rubydoc.info/gems/json_schemer/0.2.24/JSONSchemer/Errors
- https://github.com/davishmcclurg/json_schemer/blob/master/lib/json_schemer/errors.rb

### CODE REVIEW

**Comment by imjoehaines** [comment]
The messages depend on the way it failed, but basically instead of always being:

> /some/json/path failed to match /some/other/path

You can get a more useful message based on the way it failed validation, e.g.:

> property '/some/json/path' is missing required keys: a, b, c

> property '/some/json/path' is not of type: integer

Sometimes the errors are no better though, e.g. the error Karl got this morning:

> /resourceSpans/0/scopeSpans/0/spans/0/parentId failed to match /definitions/opentelemetry.proto.trace.v1.Span/additionalProperties

is equally unhelpful:

> property '/resourceSpans/0/scopeSpans/0/spans/0/parentId' is invalid: error_type=schema

**Comment by Cawllec** [comment]
> Sometimes the errors are no better though, e.g. the error Karl got this morning:

> /resourceSpans/0/scopeSpans/0/spans/0/parentId failed to match 
 /definitions/opentelemetry.proto.trace.v1.Span/additionalProperties

> is equally unhelpful:

> property '/resourceSpans/0/scopeSpans/0/spans/0/parentId' is invalid: error_type=schema

That was mostly my reasoning not to use pretty mode - while it may be slightly more helpful in some comparisons, having a path-to-path comparison will always point to the solution so long as both paths are present, which isn't always the case in pretty mode.

**Comment by imjoehaines** [comment]
> That was mostly my reasoning not to use pretty mode - while it may be slightly more helpful in some comparisons, having a path-to-path comparison will always point to the solution so long as both paths are present, which isn't always the case in pretty mode.

I don’t think:

> /resourceSpans/0/scopeSpans/0/spans/0/parentId failed to match /definitions/opentelemetry.proto.trace.v1.Span/additionalProperties

is better than:

> property ‘/resourceSpans/0/scopeSpans/0/spans/0/parentId’ is invalid: error_type=schema

Both require you to know something very unclear to determine what the problem is (`additionalProperties`/`error_type=schema`)

If you know json schemas then `additionalProperties` is kind of decipherable as it’s what determines if the object is allowed to have properties that aren’t specified in the schema — if you get an `additionalProperties` error then you can guess it means that there was an extra property and `additionalProperties` is `false`, but you have to know about json schemas first

`error_type=schema` seems similarly understandable: the json didn’t match the schema 🤷‍♀️

The other errors are way better and I think they would be more common as validation failures going forward — it’s harder to use the wrong property name than sending a badly formed value

For example, take Karl's error from this morning:

> /resourceSpans/0/scopeSpans/0/spans/0/attributes/0/value/intValue failed to match /definitions/opentelemetry.proto.common.v1.AnyValue/properties/intValue

This doesn't tell you anything until you dig into what `/definitions/opentelemetry.proto.common.v1.AnyValue/properties/intValue` is

With pretty errors it would be the much more understandable:

> "property '/resourceSpans/0/scopeSpans/0/spans/0/attributes/0/value/intValue' is not of type: string"

**Comment by Cawllec** [comment]
If you believe that pretty errors are more legible for debugging, I'm happy to go with that going forward

**Review by Cawllec — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,10 @@
   
 # 7.26.1 - 2023/04/20
 
+## Enhancements
+
+- Improve JSON schema validation failure messages [516](https://github.com/bugsnag/maze-runner/pull/516)
+
```
Comment (twometresteve) [inline CHANGELOG.md:16]: Ah, this will now need to go into the 7.27.0 section.

**Pull request merged**